chain/port/port.go

```
@@ -0,0 +1,172 @@
+	// file.
+	filePerms = 0600
+
+	// retryInterval is the interval to wait before retrying to acquire the
+	// lock file.
+	retryInterval = 10 * time.Millisecond
+
+	// maxPort is the maximum valid port number.
+	maxPort = 65535
+)
+
+var (
+	// portFileMutex is a mutex that is used to make sure that the port file
+	// is not accessed by multiple goroutines of the same process at the
+	// same time. This is used in conjunction with the lock file to make
+	// sure that the port file is not accessed by multiple processes at the
+	// same time either. So the lock file is to guard between processes and
+	// the mutex is to guard between goroutines of the same process.
+	portFileMutex sync.Mutex
+)
+
+// NextAvailablePort returns the first port that is available for listening by a
+// new node, using a lock file to make sure concurrent access for parallel tasks
+// on the same system don't re-use the same port.
+func NextAvailablePort() int {
+	portFileMutex.Lock()
+	defer portFileMutex.Unlock()
+
+	lockFile := filepath.Join(os.TempDir(), uniquePortFile+".lock")
+	lockFile = filepath.Clean(lockFile)
+	lockFileHandle := acquireLockFile(lockFile)
+
+	// Release the lock file when we're done.
+	defer func() {
+		// Always close file first, Windows won't allow us to remove it
+		// otherwise.
+		_ = lockFileHandle.Close()
+
+		err := os.Remove(lockFile)
+		if err != nil {
+			panic(fmt.Errorf("couldn't remove lock file: %w", err))
+		}
+	}()
+
+	portFile := filepath.Join(os.TempDir(), uniquePortFile)
+	portFile = filepath.Clean(portFile)
+
+	port, err := os.ReadFile(portFile)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			panic(fmt.Errorf("error reading port file: %w", err))
+		}
+
+		port = []byte(strconv.Itoa(defaultNodePort))
+	}
+
+	lastPort, err := strconv.Atoi(string(port))
+	if err != nil {
+		panic(fmt.Errorf("error parsing port: %w", err))
+	}
+
+	// We take the next one.
+	lastPort++
+	for lastPort < maxPort {
```

Looks like when the _`port file`_ reaches `65534`, the next call increments to `65535`, skips the loop (lastPort < maxPort), and panics instead of wrapping to `defaultNodePort`. Also, would need a backport for LND and unit tests, but not a serious problem asit  would take a long time to happen and only in developer environments.

---

Correct - this is mostly just a copy from `lnd`'s `port.go` - i think we could update the `lnd` to use this one instead, tho not urgent.

---

also fixed the max port case so we can readh `65535`